crates/astria-sequencer/src/accounts/ics20_transfer.rs

```diff
@@ -0,0 +1,349 @@
+        if msg.ordering != channel::Order::Unordered {
+            anyhow::bail!("channel order must be unordered for Ics20 transfer");
+        }
+
+        if msg.version_supported_on_a.as_str() != "ics20-1" {
+            anyhow::bail!("counterparty version must be ics20-1 for Ics20 transfer");
+        }
+
+        Ok(())
+    }
+
+    async fn chan_open_ack_check<S: StateRead>(_: S, msg: &MsgChannelOpenAck) -> Result<()> {
+        if msg.version_on_b.as_str() != "ics20-1" {
+            anyhow::bail!("counterparty version must be ics20-1 for Ics20 transfer");
+        }
+
+        Ok(())
+    }
+
+    async fn chan_open_confirm_check<S: StateRead>(_: S, _: &MsgChannelOpenConfirm) -> Result<()> {
+        // accept channel confirmations, port has already been validated, version has already been
+        // validated
+        Ok(())
+    }
+
+    async fn chan_close_init_check<S: StateRead>(_: S, _: &MsgChannelCloseInit) -> Result<()> {
+        anyhow::bail!("ics20 always aborts on chan_close_init");
+    }
+
+    async fn chan_close_confirm_check<S: StateRead>(
+        _: S,
+        _: &MsgChannelCloseConfirm,
+    ) -> Result<()> {
+        // no action needed
+        Ok(())
+    }
+
+    async fn recv_packet_check<S: StateRead>(_: S, _: &MsgRecvPacket) -> Result<()> {
+        // checks performed in `execute`
+        Ok(())
+    }
+
+    async fn timeout_packet_check<S: StateRead>(state: S, msg: &MsgTimeout) -> Result<()> {
+        refund_tokens_check(
+            state,
+            msg.packet.data.as_slice(),
+            &msg.packet.port_on_a,
+            &msg.packet.chan_on_a,
+        )
+        .await
+    }
+
+    async fn acknowledge_packet_check<S: StateRead>(
+        state: S,
+        msg: &MsgAcknowledgement,
+    ) -> Result<()> {
+        // see https://github.com/cosmos/ibc-go/blob/3f5b2b6632e0fa37056e5805b289a9307870ac9a/modules/core/04-channel/types/acknowledgement.go
+        // and https://github.com/cosmos/ibc-go/blob/3f5b2b6632e0fa37056e5805b289a9307870ac9a/proto/ibc/core/channel/v1/channel.proto#L155
+        // for formatting
+        let ack: TokenTransferAcknowledgement =
+            serde_json::from_slice(msg.acknowledgement.as_slice())?;
+        if ack.is_successful() {
+            return Ok(());
+        }
+
+        refund_tokens_check(
+            state,
+            msg.packet.data.as_slice(),
+            &msg.packet.port_on_a,
+            &msg.packet.chan_on_a,
+        )
+        .await
+    }
+}
+
+async fn refund_tokens_check<S: StateRead>(
+    state: S,
+    data: &[u8],
+    source_port: &PortId,
+    source_channel: &ChannelId,
+) -> Result<()> {
+    use prost::Message as _;
+
+    let packet_data = FungibleTokenPacketData::decode(data)
+        .context("failed to decode packet data into FungibleTokenPacketData")?;
+    let asset = IbcAsset::from_str(&packet_data.denom).context("invalid denomination")?;
```

Comment:
Once you have impl'ed `FromStr` you can use `str::parse`:
```suggestion
    let asset = packet_data.denom.parse::<IbcAsset>().context("failed parsing `denom` field packet data as IbcAsset")?;
```